Update to laminas-coding-standard 2.0

arueckauer opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | yes
| BC Break      | yes
| QA            | yes

### Description

This PR updates `laminas-coding-standard` to version `2.0` and provides necessary code changes.

Incompatible parts are being ignored now and need to be revisited in a later change. Most, if not all, incompatibilities are because DocBlocks instead of type declarations are used for reflection.

michalbundyra committed:
Merge branch 'docs/update' into develop

michalbundyra committed:
Merge branch 'docs/update'

michalbundyra committed:
Adds .github directory into .gitattributes

Signed-off-by: Michał Bundyra <contact@webimpress.com>

michalbundyra committed:
Require specific repository_dispatch event type - docs-build

Signed-off-by: Michał Bundyra <contact@webimpress.com>

michalbundyra committed:
Merge branch 'qa/github-workflow'

michalbundyra committed:
Merge branch 'qa/github-workflow' into develop

michalbundyra committed:
Set version constraint for replaced repository in composer.json

Signed-off-by: Michał Bundyra <contact@webimpress.com>

michalbundyra committed:
Merge branch 'hotfix/composer-replace'

michalbundyra committed:
Merge branch 'hotfix/composer-replace' into develop

michalbundyra committed:
Updates copyright notes

Signed-off-by: Michał Bundyra <contact@webimpress.com>

michalbundyra committed:
Merge branch 'copyright-update' into develop

arueckauer committed:
Update laminas-coding-standard to 2.0

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Update laminas-coding-standard to 2.0

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Update laminas-coding-standard configuration

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Update laminas-coding-standard configuration

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Apply changes required by laminas-coding-standard 2.0
Disable parts which are currently incompatible

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add interface name changes to migration guide

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add interface name changes to migration guide

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

weierophinney committed:
docs: adds CHANGELOG entry for #6

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented on composer.json:
```diff
@@ -30,7 +30,7 @@
     },
     "require-dev": {
-        "laminas/laminas-coding-standard": "~1.0.0",
+        "laminas/laminas-coding-standard": "^2.0",
```
Please use `~2.0.0`. New minor versions may introduce new sniffs... which would make CS checks fail. By using the `~` constraint, we can opt-in to those new sniffs when we are ready, instead of having to suddenly update for a new patch.

arueckauer replied:
👍🏻 Thanks for the explanation. Done.

weierophinney commented on docs/book/v3/reference/migration.md:
```diff
@@ -18,6 +18,14 @@ previously might have worked will now raise a `TypeError`.
 The following signatures changed that could affect class extension and/or
 consumers.
 
+### Interface suffix
+
+As of
+[laminas-coding-standard](https://github.com/laminas/laminas-coding-standard)
+version `2.0`, interface names must have an `Interface` suffix. The `Server`
+interface has been renamed to `ServerInterface` and the `Client` interface to
+`ClientInterface`.
+
```
We should probably introduce the `Interface` suffix variants in a new minor release on the 2.X series, and have the old, non-suffixed versions extend them. This will allow deprecation.

Ideally, we should also provide a laminas-cli command in this package to help users update any implementations they have to the new variants; that should be done in the 2.x series as well.

arueckauer replied:
The new interfaces will be introduced in 2.9.

For the command to work, one would execute the command, updating userland code to the usage of the new interfaces, right? That would mean, a minor update to make the command available and immediately upgrade to 3.x, because the new interfaces are not accepted in 2.x as arguments. To me it would make more sense to upgrade directly to 3.x and execute the command. Maybe I am missing the value of an intermediate minor update(?).

weierophinney replied:
I got it backwards: have the _new_ interfaces extend the _old_interfaces.

We did this in laminas-eventmanager and laminas-servicemanager going into v3. By introducing them in v2 releases, users were able to update their code to implement the new interfaces, while keeping backwards compatibility. Then, when they upgraded to version 3, they did not need to make any further changes.

Most of those consumers are Laminas packages. This essentially allows them to pin to **either** the 2.X series **or** the 3.X series, if they update their interfaces.

Now that we have a stable release of laminas-cli (stable in that there is a version released that you can pin to semantically in order to keep a stable API — we dropped version 0.1.0 this morning), you can even go one step further: provide a tool to help the user migrate their implementations to the new interfaces. Something like `laminas server:migrate:v3-interfaces .`, which would go through the current directory, find implementations, and update them.

Of course: all of this is moot if the v2 and v3 interfaces are not compatible. If that's the case, then scratch the whole thing, and instead just provide a command in v3 for helping migrate to the new interfaces (or have us supply [Rector configuration](https://getrector.org/) for doing that).

arueckauer replied:
> I got it backwards: have the new interfaces extend the _old_interfaces.

That would have been nice to know in #8.

arueckauer replied:
PR with changes for 2.9 is provided. I opened an issue for a possible migration command (see #11).

weierophinney commented on src/AbstractServer.php:
```diff
@@ -40,10 +40,10 @@ public function getFunctions(): Definition
      */
-    // @codingStandardsIgnoreStart
+    // phpcs:disable
     protected function _buildCallback(Reflection\AbstractFunction $reflection): Method\Callback
```
Follow the guidelines in the docblock here, and you can then get rid of the `phpcs:disable`/`phpcs:enable` dance. :wink: (make sure you also update any calls to the method)

Honestly, I'd likely add `private function buildCallback()` in v2, and have `_buildCallback()` proxy to it. Since it's an internal method, we can do this safely, and `_buildCallback()` (a) could be marked `@deprecated`, and (b) would never be called anyways, making overriding it by end-users moot. Then for 3.0, we just remove the underscore-prefixed version.

I'd do that for all underscore-prefixed methods except `_buildSignature()` and `_dispatch()` (which are the only methods extensions will consume).

That would, of course, mean having a new minor version before we merge this particular PR. But I think that it's a worthwhile exercise.

arueckauer replied:
If I haven't missed anything, everything mentioned here should be covered. Let me know if further changes are required. Resolving this for now.

weierophinney commented on src/AbstractServer.php:
```diff
@@ -67,10 +67,10 @@ protected function _buildCallback(Reflection\AbstractFunction $reflection): Meth
      * @throws Exception\RuntimeException on duplicate entry
      */
-    // @codingStandardsIgnoreStart
+    // phpcs:disable
```
I'd do two things with this method:

- Remove the underscore prefix (per the docblock!)
- Make the method `final` (so it cannot be overridden by end-users)

This will allow getting rid of the phpcs dance, and also fulfill the ideas already sketched for 3.0.

weierophinney reviewed (changes requested):
I think there are a few things we need to do in a new minor release before finalizing this patch.

arueckauer committed:
Update version constraint for laminas-coding-standard

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Update version constraint for laminas-coding-standard

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add ClientInterface and ServerInterface classes, which replaces former interfaces.
Add deprecation notice to Client and Server interfaces. They will be removed in 3.0 to match laminas-coding-standard.

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add method buildCallback()
Change _buildCallback() to proxy to buildCallbck() in preparation for removal

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Remove underscore prefix of buildSignature()
Make buildSignature() final

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Revert BC break introduced by implementing new ServerInterface

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Restore _buildSignature() as proxy method

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Revert BC break introduced by argument type declaration of newly added ServerInterface

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add remark since when class is deprecated

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add remark since when class is deprecated

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Revert BC break introduced by usage of new ServerInterface

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

weierophinney committed:
docs: adds CHANGELOG entries for #8

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
Merge branch 'develop' into dev-3.0

Conflicts:
	composer.json
	src/AbstractServer.php
	src/Client.php
	src/Server.php

weierophinney commented:
@arueckauer I've merged the changes from #8 to the dev-3.0 branch; please rebase and push again. You should be able to also remove the deprecated items on this commit! :tada:

arueckauer committed:
Resolve merge conflicts

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Resolve merge conflicts

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer commented:
@weierophinney Rebase is done and changes are provided. Ready for your review. Only pendent point is the laminas-cli command, which needs clarification (could be provided outside of this PR). Thank you.

weierophinney commented on docs/book/v3/reference/migration.md:
```diff
@@ -18,6 +18,14 @@ previously might have worked will now raise a `TypeError`.
+
+As of
+[laminas-coding-standard](https://github.com/laminas/laminas-coding-standard)
+version `2.0`, interface names must have an `Interface` suffix. The `Server`
```
I wouldn't mention that this is a change dictated by laminas-coding-standard; I'd mention it's being done for consistency with the rest of the project, where we have standardized on the `Interface` suffix.

arueckauer replied:
Done.

weierophinney commented on src/AbstractServer.php:
```diff
@@ -111,29 +100,28 @@ protected function _buildSignature(Reflection\AbstractFunction $reflection, $cla
 
         if ('function' === $type) {
             $function = $callback->getFunction();
+            // phpcs:disable
             return call_user_func_array($function, $params);
+            // phpcs:enable
```
```suggestion
            return $function(...$params);
```

We're bumping the PHP version. We can use the fact that there is now uniform callable syntax and the splat operator. :)

arueckauer replied:
Edit: Change provided to replace `call_user_func_array`.

weierophinney commented on src/AbstractServer.php:
```diff
@@ -111,29 +100,28 @@ protected function _buildSignature(Reflection\AbstractFunction $reflection, $cla
         $method = $callback->getMethod();
 
         if ('static' === $type) {
+            // phpcs:disable
             return call_user_func_array([$class, $method], $params);
+            // phpcs:enable
```
```suggestion
            $callback = [$class, $method];
            return $callback(...$params);
```

Same comment here; simplify! Doing so removes the need for ignoring CS rules.

arueckauer replied:
Edit: Change provided to replace `call_user_func_array`.

weierophinney commented on src/AbstractServer.php:
```diff
@@ -146,16 +134,13 @@ protected function _dispatch(Method\Definition $invokable, array $params)
                 $object = new $class();
             }
         }
+        // phpcs:disable
         return call_user_func_array([$object, $method], $params);
+        // phpcs:enable
```
```suggestion
        $callback = [$object, $method];
        return $callback(...$params);
```

Same comment here.

arueckauer replied:
Edit: Change provided to replace `call_user_func_array`.

weierophinney commented on src/AbstractServer.php:
```diff
@@ -146,16 +134,13 @@ protected function _dispatch(Method\Definition $invokable, array $params)
      */
-    abstract protected function _fixType(string $type): string;
-    // @codingStandardsIgnoreEnd
+    abstract protected function fixType(string $type): string;
```
We might want to document the _purpose_ of this method, for those implementing it.

arueckauer replied:
Agreed. However, this is a documentation issue which is IMO out-of-scope of this PR. I opened an issue for that.

weierophinney commented on src/ClientInterface.php:
```diff
@@ -21,5 +21,5 @@ interface Client
      * @return mixed Remote call results.
      */
-    public function call(string $method, array $params = []);
+    public function call($method, $params = []);
```
Why are you removing the parameter typehints here?

arueckauer replied:
You are right, this makes no sense. I am not sure, what introduced this. Re-adding type declarations.

weierophinney commented on src/Reflection/Node.php:
```diff
@@ -32,14 +35,16 @@ public function __construct($value, ?self $parent = null)
     }
 
+    // phpcs:disable
     /**
      * Set parent node
      *
      * @param \Laminas\Server\Reflection\Node $node
      * @param  bool $new Whether or not the child node is newly created
      * and should always be attached
      */
-    public function setParent(Node $node, bool $new = false): void
+    // phpcs:enable
```
Why are you having phpcs ignore this block?

arueckauer replied:
CS validation will fail because of the FQCN (expecting `Node`). The DocBlock param type is currently used in reflection. Change it to `Node` will fail test `ReflectionMethodTest::testMethodDocBlockFromParent`. With argument types in place, I would like to check whether it is possible to drop the DocBlock reflection in favor of using the argument types instead. I will have to do some research to see, if it will work (problematic could be mixed types).

weierophinney replied:
Ah, right - forgot that the component had not yet been updated to get type information via Reflection instead of the docblocks!

I think that will likely work really well. For mixed or union types, you still need annotations, so we can very likely fallback to parsing annotations only if the typehint is missing.

But agreed - that's for another set of changes.

weierophinney commented on src/Reflection/Node.php:
```diff
@@ -77,11 +86,17 @@ public function getParent(): ?self
         return $this->value;
     }
 
+    /**
+     * @param mixed $value
+     */
     public function setValue($value): void
```
Docblock says `return mixed`, but declaration says `void`...

arueckauer replied:
The argument is hinted with `mixed`. Return type hint in DocBlock is obsolete due to return type declaration.

weierophinney commented on test/Reflection/ReflectionFunctionTest.php:
```diff
@@ -193,12 +195,14 @@ function function2(): void
  * @return void
  */
+// phpcs:enable
 function function3(string $var1): void
```
These are largely a remnant from the days when visibility wasn't used often. I think it might make more sense to update these, add visibility, and update the docblocks, instead of ignoring phpcs.

arueckauer replied:
CS is ignored here, since DocBlock is required for reflection of the component. CS will fail validation due to redundant param and return tags. Removing DocBlock will fail test `ReflectionFunctionTest::testMultipleWhitespaceBetweenDoctagsAndTypes`.

weierophinney commented on test/TestAsset/ReflectionTestClass.php:
```diff
@@ -55,4 +56,5 @@ protected function _one(string $arg1, ?array $arg2 = null): string
     public static function two(string $arg1, string $arg2)
     {
     }
+    // phpcs:enable
```
Same comment here: update the asset to fix CS, versus ignoring CS.

arueckauer replied:
Change provided.

weierophinney commented on test/TestAsset/reflectionTestFunction.php:
```diff
@@ -23,3 +24,4 @@
 function reflectionTestFunction(bool $arg1, $arg2, string $arg3 = 'string', $arg4 = 'array')
 {
 }
+// phpcs:enable
```
And here, too.

arueckauer replied:
In the current implementation where DocBlock param tags are parsed, it is not possible to satisfy CS. This has to be reviewed with a possible change of parsing argument type declarations.

weierophinney reviewed (changes requested):
Getting closer!

In a previous comment, I've indicated what I think should happen with regards to interface changes and tooling; I'll leave the final decision to you on that.

I've flagged some additional improvements to make, mostly providing ways we can NOT ignore CS, and instead fix the code so it doesn't fail CS.

arueckauer committed:
Improve description on reasons behind renaming interfaces to a common rule on Interface suffix

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Add argument type declarations for ClientInterface::call()

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Fix CS by removing DocBlock

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Replace call_user_func_array()

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Replace call_user_func_array()

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer committed:
Replace call_user_func_array()

Signed-off-by: Andi Rückauer <arueckauer@gmail.com>

arueckauer commented:
@weierophinney Thanks for the feedback. I provided the requested changes.

There are three discussions left, all addressing the same CS issue caused by the component's usage of DocBlocks. As mentioned, I would like to investigate an improved solution by using type declarations and dropping support for DocBlocks altogether. This can be addressed after completion of the component maintenance: adding static analysis, bump to PHP 7.4 (or at least 7.3), add typed properties and introduce callables and Iterators. All these I would like to address in separate PRs.

weierophinney reviewed (approved):
:ship:

I can't remember — did I give you write access to this repo yet, @arueckauer ? If not, I can merge, but if so, go for it!

arueckauer commented:
@weierophinney You offered me becoming a maintainer, but we did not follow. For now I am happy, if you would merge it. Thank you.

weierophinney commented:
Merged - I'll get you setup as a maintainer shortly!
